pkg/controller/syncidentityprovider/syncidentityprovider_controller.go

```diff
@@ -0,0 +1,339 @@
+
+	logger log.FieldLogger
+}
+
+type identityProviderPatch struct {
+	Spec identityProviderPatchSpec `json:"spec,omitempty"`
+}
+
+type identityProviderPatchSpec struct {
+	IdentityProviders []openshiftapiv1.IdentityProvider `json:"identityProviders"`
+}
+
+// Reconcile reads that state of the cluster for a ClusterDeployment object and makes changes to the
+// remote cluster MachineSets based on the state read and the worker machines defined in
+// ClusterDeployment.Spec.Config.Machines
+// +kubebuilder:rbac:groups=hive.openshift.io,resources=clusterdeployments,verbs=get
+// +kubebuilder:rbac:groups=hive.openshift.io,resources=syncsets,verbs=get;create;update;delete;patch;list;watch
+func (r *ReconcileSyncIdentityProviders) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	contextLogger := r.logger.WithFields(log.Fields{
+		"NamespacedName": request.NamespacedName,
+	})
+
+	contextLogger.Debug("Starting Reconciliation Loop")
+
+	// Fetch the ClusterDeployment instance
+	cd := &hivev1.ClusterDeployment{}
+	err := r.Get(context.TODO(), request.NamespacedName, cd)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return
+			contextLogger.Info("cluster deployment not found")
+			return reconcile.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request
+		log.WithError(err).Error("error looking up cluster deployment")
+		return reconcile.Result{}, err
+	}
+
+	contextLogger.Data["clusterDeployment"] = cd.Name
+	contextLogger.Data["namespace"] = cd.Namespace
+
+	return reconcile.Result{}, r.syncIdentityProviders(cd)
+}
+
+func (r *ReconcileSyncIdentityProviders) newSyncSetSpec(cd *hivev1.ClusterDeployment, idps []openshiftapiv1.IdentityProvider) (*hivev1.SyncSetSpec, error) {
+	idpPatch := identityProviderPatch{
+		Spec: identityProviderPatchSpec{
+			IdentityProviders: idps,
+		},
+	}
+
+	patch, err := json.Marshal(idpPatch)
+	if err != nil {
+		return nil, fmt.Errorf("Failed marshaling identity provider list: %v", err)
+	}
+
+	return &hivev1.SyncSetSpec{
+		ClusterDeploymentRefs: []corev1.LocalObjectReference{
+			{
+				Name: cd.Name,
+			},
+		},
+		SyncSetCommonSpec: hivev1.SyncSetCommonSpec{
+			Patches: []hivev1.SyncObjectPatch{
+				{
+					APIVersion: oauthAPIVersion,
+					Kind:       oauthKind,
+					Name:       oauthObjectName,
+					PatchType:  types.MergePatchType,
+					Patch:      patch,
+				},
+			},
+		},
+	}, nil
+}
+
+func (r *ReconcileSyncIdentityProviders) syncIdentityProviders(cd *hivev1.ClusterDeployment) error {
+	idpsFromSSIDP, err := r.getRelatedSelectorSyncIdentityProviders(cd)
+	if err != nil {
+		return err
+	}
+
+	idpsFromSIDP, err := r.getRelatedSyncIdentityProviders(cd)
+	if err != nil {
+		return err
+	}
+
+	allIdps := append([]openshiftapiv1.IdentityProvider{}, idpsFromSSIDP...)
+	allIdps = append(allIdps, idpsFromSIDP...)
+
+	// Create a SyncSetSpec that includes all IdentityProviders as a patch
+	newSyncSetSpec, err := r.newSyncSetSpec(cd, allIdps)
```

Comment:
probably not a good idea to name the variable the same as the func (for maintainability)